### DISCUSSION TITLE
StringTemplate for target actions in grammars

### DISCUSSION THREAD

**Discussion by mike-lischke** [mike-lischke/antlr4-runtime-benchmarks#3] [Ideas]
> I don't use the `@members` 

Sorry, you are right. I just noticed that you already have the code in base classes.

> but still, you have to import these classes and the calls to methods are different (depending on the target)

Yes, that is true, but that's exactly what "transformGrammar is all about. Yes, it is a totally disgusting hack, but it allows for the grammar to be shared for all targets.

Rather than have all these different syntax, the grammar always uses "this." in all actions, and transformGrammmar.py replaces the pattern with the appropriate syntax. To be really "target agnostic", you shouldn't even use operators like `<`, `%`, etc. I forgot, but there is one target for one grammar in grammars-v4 that had this issue.

For Cpp, the semantic action https://github.com/mike-lischke/antlr4-runtime-benchmarks/blob/a971852628d9ecca0ce79c7972e3c026017bdadc/src/mysql/targets/antlr4-cpp/MySQLLexer.g4#L474C90-L474C111 does not have a "this->", but is it required because the field is already declared in the base class. In most grammars, the "transformGrammar.py" hack is required to mutate "this." to "this->".

For JavaScript and TypeScript, "this." must be used https://github.com/mike-lischke/antlr4-runtime-benchmarks/blob/a971852628d9ecca0ce79c7972e3c026017bdadc/src/mysql/targets/antlr4/MySQLLexer.g4#L486C90-L486C116 . But, since we already require "this." for all field and method references for "target agnostic format", it's fine as is.

For Go, lexer predicates must be reference with "l." which parser predicates are "p.".

Etc., etc., etc.

A much better solution would be to provide a StringTemplate expansion for the target. A PR was added for this here: https://github.com/antlr/antlr4/pull/4345 , but personally, I'd make the entire file a string template, and prefix or suffix the file with .st, and just apply ST through the whole file. Trash already does this. (You have to have a tool like trgen because Java requires base class code to be in a specific directory structure if you use a `-package` option for the Antlr tool.

_Originally posted by @kaby76 in https://github.com/mike-lischke/antlr4-runtime-benchmarks/discussions/1#discussioncomment-8972805_

**Comment by mike-lischke** [comment]
The problem with ST for action code is that the grammar author will have to learn that ugly syntax without any help. It might be an isolated solution for the grammar repo, but I don't think this works well for the everyday grammar writer. In fact I think ST is overused in ANTLR. Like: if you only have a hammer, every problem looks like a nail.

**Reply by kaby76** [reply]
It's hard to say whether ST actions (https://github.com/antlr/antlr4/pulls?q=is%3Apr+is%3Aopen+template) is going to be any "better" than `@members` for each target with actions. As they are set up with the PR, they're basically macros. Either way, it's makes sense to not make lots of copies of the .g4's.

I don't think there will be another release of Antlr4 4.13. And, the PR was never merged because the developer didn't address all the concerns.
